common/hashing.h

```diff
@@ -0,0 +1,721 @@
+//
+// This type is primarily used by types to implement a customization point
+// `CarbonHash` that will in turn be used by the `HashValue` function. See the
+// `HashValue` function for details of that extension point.
+//
+// The API of this type can be used to incorporate data from your user-defined
+// type into a hash code. The API is a low-level collection of primitives rather
+// than a higher-level abstraction. This reflects a fundamental priority: deep,
+// detailed, and latency-oriented optimization of any hash functions. This
+// type's API also reflects the reality that high-performance hash tables rely
+// on keys that are generally small and cheap to hash. The result is
+// prioritizing hashing a small number of integers (1 or 2 ideally), or some
+// contiguous byte buffer. We typically want hash tables whose key types work to
+// fit into this model rather than advanced or more complex hashing of more
+// complex input data.
+//
+// It also results in an unusual API design where all the functions are actually
+// *static* member functions that expect the `Hasher` and its state to be moved
+// into the first parameter and returned, updated, by value. This is in contrast
+// to non-static member functions in C++ which would mutate the state in place
+// by taking its address for the implicit `this` parameter. While that *usually*
+// optimizes away, it creates a more difficult optimization environment that has
+// resulted in performance issues over time and at scale.
+//
+// Another important note is that types only need to include the data that would
+// lead to an object comparing equal or not-equal to some other object,
+// including objects of other types if a type supports heterogeneous equality
+// comparison.
+//
+// To illustrate this -- if a type has some fixed amount of data, maybe 32
+// 4-byte integers, and equality comparison only operates on the *values* of
+// those integers, then the size (32 integers, or 128 bytes) doesn't need to be
+// included in the hash. But if a type has a *dynamic* amount of data, and the
+// sizes are compared as part of equality comparison, then that dynamic size
+// *does* need to be included in the hash. An interesting observation of this
+// principle is that C-strings don't need to incorporate their length into their
+// hash *if* they incorporate the null terminating byte in the hash -- because
+// that terminating byte will be the difference in equality comparison. But if
+// the C-strings can be compared with any kind of explicitly-sized string views,
+// the size *does* need to be incorporated.
+class Hasher {
+ public:
+  Hasher(Hasher&& arg) = default;
+  Hasher(const Hasher& arg) = delete;
+  auto operator=(Hasher&& rhs) -> Hasher& = default;
+
+  // Extracts the current state as a `HashCode` for use.
+  explicit operator HashCode() const { return HashCode(buffer); }
+
+  // Convenience method to hash an object by its object representation when that
```

Comment:
This very much does have a contract, it's just kind of indirect. Here's how I'd recommend approaching this (drawn largely from [P0029](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0029r0.html)):

- `Hasher`'s internal state is notionally an arbitrary-length sequence of bytes.
- These methods all append a specified sequence of bytes to that internal state (e.g. this one appends the object representation).
- `HashValue` computes a `HashCode` from the internal state in an unspecified way.
- The bytes that a `CarbonHash` overload appends to the hasher's internal state are called the _hash representation_ of the value, and they are required to have the following properties:
   - If two objects of the same type compare equal, their hash representations will be equal.
   - If two objects of the same type do not compare equal, with high probability (ideally probability 1) their hash representations will be different.
   - If two objects of the same type do not compare equal, with high probability (ideally probability 1) neither hash representation will be a suffix of the other.

That third condition is why we need to treat the `Hasher` state as an ordered sequence rather than the more amorphous phrasing you currently have. See my comment above for why we need that third condition at all.
(Technically the third condition subsumes the second one, but it's probably better to spell the second one out for clarity.)